### ISSUE TITLE
EditorPlugins that depend on the resources that Node has will not be visible when SubInspector is opened

### ISSUE THREAD

**Issue by TokageItLab** [godotengine/godot#70914]
### Godot version

4.0.dev(dc3175e4cdb4e54e1c6200b6c4d07646eea95284)

### System information

Any

### Issue description

EditorPlugins that depend on the resources that Node has will not be visible when SubInspector is focused.

https://user-images.githubusercontent.com/61938263/210553656-86669f57-e41f-427e-a650-30630c27d850.mp4

This is caused by two problems:

The first is caused by implicit editor deletion in Node3D. This is fixed by #70907.

The other is due to a change in inspector update timing, which is regression by #68498. Revert PR #70909 will fix this issue, but if not, we will need to update SubInspectorEditor in the correct way.

### Steps to reproduce

Select Path3D or CollisionPolygon3D and open a SubInspector such as Script.

### Minimal reproduction project

[plugintest.zip](https://github.com/godotengine/godot/files/10343962/plugintest.zip)

**Comment by YuriSizov** [comment]
What is supposed happen when you click on the script resource?

**Comment by TokageItLab** [comment]
@YuriSizov 

> What is supposed happen when you click on the script resource?

Nothing happen and it keeps bugging, the Curve/PolygonEditorPlugin in the viewport is hidden.

> The path should stay editable in the viewport?

Yes, the SubInspector is only open rather than being focused, and the bug is still happening the moment I select the Node, not just the moment I set the Script.

Perhaps this is because “the process of keeping SubInspector open“ removes the EditorPlugin that depends on other inspector resources. It looks like there may be a fundamental problem somewhere in the Inspector's update method, but I don't know.

**Comment by YuriSizov** [comment]
This doesn't seem to have anything to do with the update tree changes, other than... it is possible that the old and excessive update tree behavior would self-correct the issue, unintentionally. But otherwise this just looks like our old friend "multiple editor plugins cannot be active at once".

**Comment by YuriSizov** [comment]
> I don't think this is a truth now.

Well, we never changed anything about it, so it's still true. KoBeWi had a plan to add support for multiple plugins to be active at once, but reduz didn't want the complexity and instead it was fixed for the majority of offenders on the case-by-case basis. That's why we have the shader editor always on now. But still many plugins have hacks to work around this limitations.

> In practice, this works correctly when you revert the update changes by #70909.

As mentioned, I see no evidence that the changes in that PR had anything to do with your problem, other than perhaps rapid updates were overriding the state of the inspector hiding the issue with these few plugins that are now exposed.

**Comment by Rindbee** [comment]
I can't reproduce it in beta8 and master, clicking on the script property in the Inspector always jumps to the script editor, and sets the edited object to that script in the Inspector, just same as disabling `interface/inspector/open_resources_in_current_inspector `.

Need to modify some settings?

**Comment by Rindbee** [comment]
> It seems to jump to ScriptEditor if open_resources_in_current_inspector is disabled. It needs to be enabled.

https://user-images.githubusercontent.com/30386067/210568353-6edb44d4-3db1-47dd-9f5b-2090d35685a2.mp4

I deleted the cache dir, config dir, and data dir, and tried again, and it was the same as before.

**Comment by YuriSizov** [comment]
@Rindbee There is a setting to stop switching to the script editor, I think. But regardless, the issue is reproduceable, I can confirm this.

**Comment by TokageItLab** [comment]
@Rindbee Maybe it needs to remove 'Script' from `Resource to Open in New Inspector`.

**Comment by TokageItLab** [comment]
Probably not completely fixed. At least, Skeleton3DEditor is fixed by #70909 and #70907, but the CurveEditor and CollisionPolygonEditor are still broken by script's SubInspector.

**Comment by YuriSizov** [comment]
> At least, Skeleton3DEditor is fixed by #70909 and #70907, but the CurveEditor and CollisionPolygonEditor are still broken by script's SubInspector.

They get into a bad state by any new edited object: you can click on the Curve resource for the Path node and have the same issue. And you can reproduce it in beta 1 (though not in 3.x, so it is some kind of regression technically).

This all points to the fact that the revert of the optimization wasn't really necessary. The bug is not related to it. The main improvement from the revert is that deselecting the node and then reselecting it now fixes the state of the editor. But folding and unfolding any resource breaks it again.

Long story short: it's the specific plugins that need to be fixed.

**Comment by Rindbee** [comment]
The `update_tree` call should not affect the plug-ins used for nodes, at most it will only affect the plug-ins used for resources.

**Comment by YuriSizov** [comment]
> The `update_tree` call should not affect the plug-ins used for nodes, at most it will only affect the plug-ins used for resources. Of course there may be some exceptions.

Indeed, these issues have nothing to do with `update_tree`. Updating the tree is purely a process of recreating the inspector. These issues are about plugins not handling edited objects correctly.

**Comment by TokageItLab** [comment]
Indeed, the SkeletonEditor bug is fixed by https://github.com/godotengine/godot/pull/70907 only; https://github.com/godotengine/godot/pull/70909 seems to be irrelevant.

Also, I agree that the optimization only mainly causes problems with the Curve resource in the Curve3DEditor (although there was a problem with AnimationPlayerEditor), but that there are bugs independent of the optimization.

If we fix the plugin so that the above script's SubInspector bug does not occur, it is quite possible that the optimization will not cause the bug.

However, I want to avoid clear bugs as we are close to RC for the time being.

So for now, we should fix the plugin before optimization. After that, we recommend reapplying the optimization, taking note of the problem that occurred this time.

**Comment by YuriSizov** [comment]
Oh, Path3D editor plugin is a convoluted mess. So what happens when you select the script in the Inspector is the following:

1. Script resources have an editor plugin attached to them, `ScriptEditorPlugin`, which is a plugin that has a top bar menu; the normal logic for this is to activate the top bar menu to switch to that plugin's view, and naturally hide all other plugins that may be currently visible. For this `make_visible` is called on all of them, with `false` being passed to other plugins, including `Path3DEditorPlugin`.
2. `Path3DEditorPlugin` handles this call by hiding parts of the UI it adds (the toolbar), but also by clearing its stored `Path3D` reference, and sending a `changed` signal to its `Curve3D` resource.
3. When a `Curve3D` resource attached to a `Path3D` node changes, it is handled by the `Path3D::_curve_changed` method, which in turn calls into `Node3D::update_gizmos`, inherited `Node3D`. This method is editor-only and is responsible for updating the gizmos, naturally. In this case it adds to the MessageQueue a request to call its internal counterpart that does the actual updating, `Node3D::_update_gizmo`.
4. The internal method checks for simple things, it makes sure that the node is still visible in the tree (which it is) and then calls for a `redraw` on the gizmo object. If the node was not visible, it would call `clear`. But this time we trigger a redraw.
5. This brings us back to the `Path3DEditorPlugin`, or rather its auxiliary class `Path3DGizmo`. `Path3DGizmo::redraw` does the normal gizmo object thing and constructs visual elements for the gizmo. This includes a visualization of the curve, as well as handles for controlling said curve. Problem is, handles are only supposed to be drawn for the currently edited path. If we go back to point 2 of this list, we will remind ourselves that we just cleared the edited path. So of course we decide that there is a mismatch and we don't have anything to draw. Thus handles disappear.

Now, you might wonder why we call `make_visible` on `Path3DEditorPlugin` in the first place? We never switch to the script editor, so this seems pointless. And I agree. This is pointless, and something that I'm looking into next. At the same time, the handling of `make_visible` in `Path3DEditorPlugin` is still questionable. Even if we do switch to the script editor, we probably shouldn't remove the reference of the last edited path from `Path3DEditorPlugin`. Because when we switch back to the 3D view, the gizmos will be gone anyway as a result. This is in fact what happens in 3.x, so the issue with the plugin dates way back.

So, what should we do? I'll see if we can prevent pointless `make_visible` calls, which would _technically_ fix the regression. But there is still a buggy behavior in `Path3DEditorPlugin` that needs fixing by someone who knows more about this plugin than I, it's just consistent with a buggy behavior in 3.x.

Oh, and collision polygons are kind of the same, kind of worse too. They don't have a dedicated plugin. Instead `Polygon3DEditorPlugin` is used for the polygon part. And `Polygon3DEditorPlugin` doesn't have its own gizmos. Instead it spawns a mesh instance that has gizmos. Upon receiving `make_visible(false)` it resets its inner state (by calling `edit(nullptr)`), effectively removing the mesh instance and all of its bells and whistles.

**Comment by TokageItLab** [comment]
I guess there is a hint as to why Skeleton3DEditor is working. Skeleton3DEditor now uses SubGizmo, so the place where the drawing is performed may be different from those EditorPlugin. But my memory is a bit hazy...

**Comment by YuriSizov** [comment]
@TokageItLab `Skeleton3DEditor` isn't an `EditorPlugin`, so doesn't appear to implement `make_visible` at all. It's just a UI element of the inspector plugin, `EditorInspectorPluginSkeleton`. Their owner, `Skeleton3DEditorPlugin`, doesn't implement `make_visible` either.

So yeah, as long the edited object is visible in the inspector, its gizmos should be displayed correctly in this case. Plugins for `Path3D` and `CollisionPolygon3D` work independently from the inspector. Maybe that's how they should be rewritten then, no new system needed. Just use subgizmos. Should be pretty easy for `Path3D`, but not for `CollisionPolygon3D` due to that intermediate mesh instance.

**Comment by YuriSizov** [comment]
Okay, I don't think we can fix the issue with clicking on the script resources making gizmos disappear right now. As I mentioned before, this issue also exists in 3.x, you just always switch your view to the script editor. In 4.0 you can remain in the 3D view, but that doesn't really matter much to the issue itself, as the behavior is essentially the same.

Godot editor has a stack of edited objects, and the one selected last is the one that controls which plugins are active, and which are not. Selecting a script, a resource that has a main screen plugin associated with it, makes it the latest selected object, which tells all plugins that handle other objects to hide themselves. This logic is very core to the whole editing and inspecting experience, and I wouldn't want to touch it right now. We should look into improving it in future, but for now I'd leave it as is.

The situation can be improved, as mentioned above, by fixing the plugins themselves, so they don't remove the reference on being hidden, or by using subgizmos.
